### PR TITLE
Internalize Stapler so we can move away from core releases

### PR DESCRIPTION
**Pull request by i386**
# Description

PTAL @vivek 

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@reviewbybees

### CODE REVIEW

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/stapler/export/Export.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
+package io.jenkins.blueocean.commons.stapler.export;
```
Comment (vivek) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/stapler/export/Export.java]: I guess Export is our own code, not moved in from Stapler. We should keep it separate from rest of Stapler classes. For example once core allows plugin specific stapler jar and we remove exported package and change Export.doJson() without affecting other parts of the code. So maybe move it to `io.jenkins.blueocean.commons.stapler`?

File: blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/PagedResponse.java [hunk 2]
```diff
@@ -21,7 +21,7 @@
 /**
  * Renders {@link Pageable} to HTTP by honoring the current page, links to next page, etc.
  *
- * @author Kohsuke Kawaguchi
+ * @author Kohsuke KawaguchiF
```
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/PagedResponse.java]: typo:)

**Comment by vivek** [comment]
@i386 code wise looks good to me, will do some testing 👍 

I think we should have a script to copy code from stapler exported package in future, for example any bug fixes in stapler exported package, we want to bring in. Such as https://github.com/stapler/stapler/pull/89.

Some concerns/observation:

- As it sounds from the PR title, it will fix issue of upgrading core for stapler issues is not completely true. Yes it will avoid upgrade for json serialization functionality but not for the rest of stapler functionalities. So any fixes in stapler web framework would still require upgrade. 
- As discussed, its temporary measure - that is till core provides a way to use plugin mandated stapler version without upgrade of core, we keep this solution around.

**Comment by vivek** [comment]
@i386 `mvn clean install` fails due to javadoc generation error. Following patch in blueocean-commons/pom.xml fixes it (exclude exported package from generation):

```
diff --git a/blueocean-commons/pom.xml b/blueocean-commons/pom.xml
index 833331d..10df76d 100644
--- a/blueocean-commons/pom.xml
+++ b/blueocean-commons/pom.xml
@@ -21,4 +21,17 @@
       <version>${jackson.version}</version>
     </dependency>
   </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-javadoc-plugin</artifactId>
+        <version>2.10.4</version>
+        <configuration>
+          <excludePackageNames>io.jenkins.blueocean.commons.stapler.export</excludePackageNames>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
 </project>
```

Overall looks good! I completed with my testing. Tree api works fine for both pipelines and runs API. ATH passes. Pending this pom changes and some comments/concerns posted in earlier comment, LGTM 🐝

**Comment by i386** [comment]
@vivek awesome - glad you approve. Doing it with a script won't add much time to upgrading export as there are a few APIs I've had to hand delete/change for compatibility. I've also ditched all the python, ruby and xml variants to slim this down

**Comment by i386** [comment]
@vivek I also copied over the `MethodHandle` optimisations so we will have those in the next beta too

**Comment by vivek** [comment]
@i386 I left other comment on the code, looks like you missed them:) also, I guess you are going to pull in Jesse fix for WSOD improvement, once it's merged, in later PR?

**Comment by i386** [comment]
@vivek oops I'll come back to it this afternoon when I get off my flight.

I'll move over the WSOD changes when they land.

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/stapler/export/Export.java (hunk 1 above)
Comment (i386) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/stapler/export/Export.java]: Can do

**Comment by i386** [comment]
@vivek Fixed in https://github.com/jenkinsci/blueocean-plugin/commit/f04880d6018e07197b34c8194b75ff6563683703 and https://github.com/jenkinsci/blueocean-plugin/commit/9fc12c4fd58f170ef9aa17f137c74c4ffc870980

**Comment by vivek** [comment]
@i386 thanks 👍

**Pull request merged**